kafka/src/main/java/org/apache/calcite/adapter/kafka/KafkaStreamTable.java

```diff
@@ -77,14 +90,61 @@
         if (tableOptions.getConsumerParams() != null) {
           consumerConfig.putAll(tableOptions.getConsumerParams());
         }
-        Consumer consumer = new KafkaConsumer<>(consumerConfig);
-        consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()));
-
+        Consumer consumer;
+        if (tableOptions.getConsumer() != null) {
+          consumer = tableOptions.getConsumer();
+        } else {
+          consumer = new KafkaConsumer<>(consumerConfig);
+        }
+        if (tableOptions.getTimestamp() != -1) {
+          // seek to special timestamp
+          consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()),
+              new ConsumerRebalanceListener() {
+                @Override public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                }
+
+                @Override public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                  // consumer can seek to special offset after the consumer join the consumer group
+                  seekToTimestamp(consumer, tableOptions.getTopicName(),
+                      tableOptions.getTimestamp());
+                }
+              });
+        } else {
+          consumer.subscribe(Collections.singletonList(tableOptions.getTopicName()));
+        }
         return new KafkaMessageEnumerator(consumer, tableOptions.getRowConverter(), cancelFlag);
       }
     };
   }
 
+  /**
+   * Seek to special timestamp.
+   *
+   * @param consumer  KafkaConsumer;
+   * @param topicName topic to subscribe;
+   * @param timestamp the timestamp need to seek;
+   */
+  private void seekToTimestamp(Consumer consumer, String topicName, long timestamp) {
+    if (hasSeekToTimestamp) {
```

Comment:
The method is not thread-safe, and KafkaConsumer itself is not safe for multi-threaded access.
And I think we do not need to consider the case of multiple-threads, because KafkaAdapter does not support multi-threaded.